tlog-witness.md

```diff
@@ -0,0 +1,226 @@
+
+A witness MAY use the same value for both the *submission prefix* and the
+*monitoring prefix*.
+
+If exposing the machine holding the witness key material directly to the
+Internet is undesirable, operators MAY use a [bastion][].
+
+### add-checkpoint
+
+The `add-checkpoint` call is used to submit a new checkpoint to the witness,
+along with a consistency proof from a previous checkpoint, and returns the
+cosignature.
+
+    POST <submission prefix>/add-checkpoint
+
+The request MUST be an HTTP POST. Clients SHOULD use and witnesses SHOULD
+support keep-alive HTTP connections to reduce latency and load due to connection
+establishment.
+
+The request body MUST be a sequence of
+  - a previous size line,
+  - zero or more consistency proof lines,
+  - and an empty line,
+  - followed by a [checkpoint][].
+
+Each line MUST terminate in a newline character (U+000A).
+
+The previous size line MUST consist of the string `old`, a single space (0x20),
+and the tree size of the previous checkpoint encoded as an ASCII decimal with no
+leading zeroes (unless the size is zero, in which case the encoding MUST be `0`).
+
+Each consistency proof line MUST encode a single hash in base64. The client MUST
+NOT send more than 63 consistency proof lines.
+
+The submitted checkpoint MAY include multiple signatures. 
+
+    old 20852014
+    PlRNCrwHpqhGrupue0L7gxbjbMiKA9temvuZZDDpkaw=
+    jrJZDmY8Y7SyJE0MWLpLozkIVMSMZcD5kvuKxPC3swk=
+    5+pKlUdi2LeF/BcMHBn+Ku6yhPGNCswZZD1X/6QgPd8=
+    /6WVhPs2CwSsb5rYBH5cjHV/wSmA79abXAwhXw3Kj/0=
+
+    example.com/behind-the-sofa
+    20852163
+    CsUYapGGPo4dkMgIAUqom/Xajj7h2fB2MPA3j2jxq2I=
+
+    — example.com/behind-the-sofa Az3grlgtzPICa5OS8npVmf1Myq/5IZniMp+ZJurmRDeOoRDe4URYN7u5/Zhcyv2q1gGzGku9nTo+zyWE+xeMcTOAYQ8=
+    — example.com/behind-the-sofa opLqBQsREYCgu6xQkYQwJr9fo45a62DN9EdmHXnZdXNqlcVGlCum2Wks+49/V6267UEjw6QUXTS5Rovnzv++qbSzm9Q=
+
+The witness MUST verify the checkpoint signature against the public key(s) it
+trusts for the checkpoint origin, and it MUST ignore signature from unknown
+keys. If the checkpoint origin is unknown, the witness MUST respond with a "404
+Not Found" HTTP status code. If none of the signatures verify against a trusted
+public key, the witness MUST respond with a "403 Forbidden" HTTP status code.
+
+The previous size MUST be equal to or lower than the checkpoint size. Otherwise,
+the witness MUST respond with a "400 Bad Request" HTTP status code.
+
+The witness MUST check that the previous size matches the size of the latest
+checkpoint it cosigned for the checkpoint's origin (or zero if it never cosigned
+a checkpoint for that origin). If it doesn't match, the witness MUST respond
+with a "409 Conflict" HTTP status code. The response body MUST consist of the
+tree size of the latest cosigned checkpoint in decimal, followed by a newline
+(U+000A). The response MUST have a `Content-Type` of `text/x.tlog.size`.
```

Comment:
The approach we took here is for the witness to simply return the old checkpoint back to the caller. Any reason to prefer this approach with custom headers?

Reading between the lines it seems like you're intentionally trying to avoid having the witness serve checkpoints up so that it doesn't become load-bearing on the read path?

Reply:
+1 to returning the last known checkpoint. This is helpful to avoid more roundtrips, especially for logs with significant amounts of traffic where checkpoints may be frequently updated. If a log requests a cosignature for a checkpoint that's older than what it's already verified, the cosigned checkpoint can be returned to the log and the log can verify consistency from its older checkpoint to the latest cosigned checkpoint.

Reply:
The custom headers are just a way to distinguish the witness protocol error from, say, a CDN error. I think we should have them whatever the body is.

Indeed, we want to make sure this API doesn't get abused to read checkpoints from witnesses.

How can a log request a cosignature for an older checkpoint, when the log produces the checkpoints?

Anyway, the witness certainly can't return a consistency proof, so the client here needs to have access to the tree to check any kind of consistency, so I don't think returning the whole checkpoint adds anything over returning just the size.

Reply:
> How can a log request a cosignature for an older checkpoint, when the log produces the checkpoints?

If a log is producing checkpoints per-entry rather than batching.

* Two requests to the log for entry inclusion
* The log generates a checkpoint for the first entry (size N)
* The log generates a checkpoint (size N+1) for the second entry
* The log sends two requests to the witness for N and N+1
* Checkpoint N+1 succeeds first, and the cosigned checkpoint is returned to the log
* Checkpoint N fails to be cosigned since N < N+1

> Anyway, the witness certainly can't return a consistency proof, so the client here needs to have access to the tree to check any kind of consistency, so I don't think returning the whole checkpoint adds anything over returning just the size.

Is the expectation that the log persists the cosigned checkpoint? Otherwise, if you return only the size, the log can compute the relevant checkpoint but needs to request another cosignature. If so, can we document that?

Reply:
> If a log is producing checkpoints per-entry rather than batching.

Interesting, didn't think of that!

I'd expect such a log to keep around the latest cosigned checkpoint somewhere that's accessible to the first request (since that request needs to access state from "the future" anyway to build the proofs against the latest checkpoint).

This log design gets unmanageable fast with multiple witnesses though. Even if we were to return a full cosigned checkpoint on a 409, the checkpoints returned by different witnesses could get out of sync, and then the cosignatures can't be merged.

Reply:
I think to summarize returning only the size discourages using the write path to obtain checkpoints, and the scenarios where it would be a problem are already problematic for other reasons.

> While this might be more in scope for a spec on offline signatures, do you think the spec should have any suggestions about atomic operations for the log, not just the witness?

Yeah I think that will be down to implementation preference, but I think that the conclusion that there needs to either be leader election or single-node architecture is unavoidable.